src/QsCompiler/CSharpGeneration/EntryPoint.fs

```diff
@@ -137,17 +136,113 @@ let private callableTypeNames context (callable: QsCallable) =
 let private entryPointClassFullName (entryPoint: QsCallable) =
     { Namespace = entryPoint.FullName.Namespace; Name = entryPointClassName + entryPoint.FullName.Name }
 
+/// The QIR argument type for the Q# type, or None if the Q# type is not supported in a QIR entry point.
+let rec private qirArgumentType (type_: ResolvedType) =
+    let case name =
+        "global::Microsoft.Quantum.Runtime.ArgumentType." + name |> ident
+
+    match type_.Resolution with
+    | Bool -> case "Bool" :> ExpressionSyntax |> Some
+    | Int -> case "Int" :> ExpressionSyntax |> Some
+    | Double -> case "Double" :> ExpressionSyntax |> Some
+    | Pauli -> case "Pauli" :> ExpressionSyntax |> Some
+    | Range -> case "Range" :> ExpressionSyntax |> Some
+    | Result -> case "Result" :> ExpressionSyntax |> Some
+    | String -> case "String" :> ExpressionSyntax |> Some
+    | ArrayType itemType ->
+        qirArgumentType itemType
+        |> Option.map (fun itemType -> ``new`` (case "Array") ``(`` [ itemType ] ``)``)
+    | _ -> None
+
+/// The QIR argument value for the Q# type and value expression, or None if the Q# type is not supported in a QIR entry
+/// point.
+let rec private qirArgumentValue (type_: ResolvedType) (value: ExpressionSyntax) =
+    let argumentValueName = "global::Microsoft.Quantum.Runtime.ArgumentValue"
+
+    let case name =
+        sprintf "%s.%s" argumentValueName name |> ident
+
+    let arrayValue itemValue itemType =
+        let values =
+            ident "global::System.Linq.Enumerable"
+            <.> (ident "Select", [ value; upcast ``() =>`` [ "item" ] itemValue ])
+
+        let items =
+            ident "global::System.Collections.Immutable.ImmutableArray"
+            <.> (sprintf "CreateRange<%s>" argumentValueName |> ident, [ values ])
+
+        case "Array" <.> (ident "TryCreate", [ items; itemType ])
+
+    match type_.Resolution with
```

Comment:
The duplication in what cases are and aren't supported for arguments is bothering me a bit. It could be avoided if a helper function was factored out.

Reply:
done